openapi/openapi.yml

```diff
@@ -0,0 +1,246 @@
+          # max number of rooms has been reached?
+    get:
+      tags:
+        - room
+      summary: List all rooms and their configs
+      operationId: getRooms
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Room'
+  /room/{roomId}:
+    get:
+      tags:
+        - room
+      summary: Find room by ID
+      operationId: getRoomById
+      parameters:
+        - name: roomId
+          in: path
+          description: ID of room to return
+          required: true
+          schema:
+            type: integer
+            format: int64
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Room'
+        '400':
+          description: Invalid ID supplied
+        '404':
+          description: Room not found
+    delete:
+      tags:
+        - room
+      summary: Delete a room
+      operationId: deleteRoom
+      parameters:
+        - name: roomId
+          in: path
+          description: Room id to delete
+          required: true
+          schema:
+            type: integer
+            format: int64
+      responses:
+        '400':
+          description: Invalid room value
+  /room/{roomId}/peer:
+    post:
+      tags:
+        - room
+      summary: Add peer
+      operationId: addPeer
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/PeerConfig'
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Peer'
+        '500':
+          description: reached max number of peers
+  /room/{roomId}/consumer:
+    post:
+      tags:
+        - room
+      summary: Create consumer
+      operationId: createConsumer
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ConsumerConfig'
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Consumer'
+  /room/{roomId}/producer:
+    post:
+      tags:
+        - room
+      summary: Create producer
+      operationId: createProducer
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ProducerConfig'
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Producer'
```

Comment:
These two endpoints feel like basically the same thing and a little too low-level for the high-level API that we hoped for. I think I would favor a simple `endpoint` here instead.

And I don't really see the benefit in keeping separating consumers and producers on this level. Is it permissions? Endpoints, as far as I can see, are not peers and would be a well-defined and trusted piece of code :thinking: 

I also have another problem here - this doesn't facilitate an endpoint that is both producer and consumer. Of course, we don't plan to provide it, but if someone decided to add it, this API makes it rather tricky.

Reply:
Hmm :thinking: 

> These two endpoints feel like basically the same thing and a little too low-level for the high-level API that we hoped for

I don't see any difference between having two different HTTP endpoints and one endpoint where you have to specify what kind of given engine endpoint type you want e.g. (hls_producer, hls_consumer). There are still terms `producer` `consumer` 

> I also have another problem here - this doesn't facilitate an endpoint that is both producer and consumer. Of course, we don't plan to provide it, but if someone decided to add it, this API makes it rather tricky.

The same problem is with one HTTP endpoint. How would you add an HLS engine endpoint that is both producer and consumer? It has to have some name. When we find a name for it we can create third HTTP endpoint with similar generic name

The main benefit of having separate endpoints for producers and consumers is that it is easier to distinguish from them. There is a separate enum for producers and a separate enum for consumers so it's easier to find out what producers and what consumers there are. They are also split when returning room description.

I am curious what other folks think

Reply:
> The same problem is with one HTTP endpoint. How would you add an HLS engine endpoint that is both producer and consumer? It has to have some name. When we find a name for it we can create a third HTTP endpoint with a similar generic name

1. Engine endpoints don't even have a concept of being a `producer` or `consumer`, they are what you make it - so this instantly doesn't fit the deal.
2. You will need to write basically duplicate code in the handler for these two (or three!!!) routes, or do some path parameter magic instead of writing a simple endpoint
3. Any SDK for Jellyfish will either end up with 2 functions for adding endpoints or with just one, working around that quirk.
4. It's really weird and potentially confusing. I just spoke with Paweł Kruczkiewicz about that and I found out that for him, the current HLS endpoint should count as a producer because it produces a HLS stream... So, this is not so obvious. I think this is a reason enough to use one endpoint and distinguish it with "type".

The way I see it, this very much complicates a very basic action of adding a server-side endpoint